From 0361c0cfb6f8cfd7a673b081d794daf3f451ae7c Mon Sep 17 00:00:00 2001
From: Bogdan Hamciuc <bogdan.hamciuc@nxp.com>
Date: Tue, 23 Nov 2021 11:42:37 +0200
Subject: [PATCH 08/14] plat: s32g: Reduce fip size to account for M7
 bootloader

In an integrated setup, we might have an M7 bootloader at the end of SRAM.
We reserve 1MB for it. This patch shrinks the fip size to 3MB and
preserves the offsets in SRAM of fip/bl2/dtb blobs.

Upstream-Status: Pending 

Signed-off-by: Bogdan Hamciuc <bogdan.hamciuc@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
Issue: ALB-8226
---
 plat/nxp/s32/s32g/include/s32g_platform_def.h |  2 +-
 plat/nxp/s32/s32g/s32g_common.mk              | 10 +++++++++-
 2 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/plat/nxp/s32/s32g/include/s32g_platform_def.h b/plat/nxp/s32/s32g/include/s32g_platform_def.h
index 560b7f3ea..83ed2c1e6 100644
--- a/plat/nxp/s32/s32g/include/s32g_platform_def.h
+++ b/plat/nxp/s32/s32g/include/s32g_platform_def.h
@@ -187,7 +187,7 @@
 #define S32G_BL32_BASE		(BL31_BASE - S32G_BL32_SIZE)
 #define S32G_BL32_LIMIT		(BL31_BASE)
 
-#define FIP_BASE		(S32G_SRAM_END - FIP_MAXIMUM_SIZE)
+#define FIP_BASE		(S32G_SRAM_END - FIP_ROFFSET)
 
 /* FIXME value randomly chosen; should probably be revisited */
 #define PLATFORM_STACK_SIZE		0x4000
diff --git a/plat/nxp/s32/s32g/s32g_common.mk b/plat/nxp/s32/s32g/s32g_common.mk
index df4499cf4..9dd10be1a 100644
--- a/plat/nxp/s32/s32g/s32g_common.mk
+++ b/plat/nxp/s32/s32g/s32g_common.mk
@@ -174,8 +174,16 @@ ifdef FIP_MEM_OFFSET
 $(eval $(call add_define,FIP_MEM_OFFSET))
 endif
 
-FIP_MAXIMUM_SIZE	:= 0x400000
+# Reserve some space at the end of SRAM for external apps and include it
+# in the calculation of FIP_BASE address.
+EXT_APP_SIZE		:= 0x100000
+$(eval $(call add_define,EXT_APP_SIZE))
+FIP_MAXIMUM_SIZE	:= 0x300000
 $(eval $(call add_define,FIP_MAXIMUM_SIZE))
+# FIP offset from the far end of SRAM; leave it to the C code to perform
+# the arithmetic
+FIP_ROFFSET		:= "(EXT_APP_SIZE + FIP_MAXIMUM_SIZE)"
+$(eval $(call add_define,FIP_ROFFSET))
 
 FIP_ALIGN := 512
 all: add_to_fip
-- 
2.17.1

